pkg/controllers/workgenerator/reconciler.go

```diff
@@ -0,0 +1,299 @@
+)
+
+// Reconciler watches binding objects and generate work objects in the designated cluster namespace
+// according to the information in the binding objects.
+// TODO: incorporate an overriding policy if one exists
+type Reconciler struct {
+	client.Client
+	recorder record.EventRecorder
+}
+
+// Reconcile triggers a single binding reconcile round.
+func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	klog.V(2).InfoS("Start to reconcile a ClusterResourceBinding", "cluster resource binding", req.Name)
+	startTime := time.Now()
+	bindingRef := klog.KRef(req.Namespace, req.Name)
+	// add latency log
+	defer func() {
+		klog.V(2).InfoS("ClusterResourceBinding reconciliation loop ends", "resource binding", bindingRef, "latency", time.Since(startTime).Milliseconds())
+	}()
+	var resourceBinding fleetv1beta1.ClusterResourceBinding
+	if err := r.Client.Get(ctx, req.NamespacedName, &resourceBinding); err != nil {
+		if errors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+		klog.ErrorS(err, "Failed to get the resource binding", "resource binding", bindingRef)
+		return ctrl.Result{}, controller.NewAPIServerError(true, err)
+	}
+
+	// handle the case the binding is deleting
+	if resourceBinding.DeletionTimestamp != nil {
+		return r.handleDelete(ctx, resourceBinding.DeepCopy())
+	}
+
+	// update the resource binding obj with finalizer
+	if err := r.ensureFinalizer(ctx, &resourceBinding); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// we only care about the bound bindings
+	if resourceBinding.Spec.State != fleetv1beta1.BindingStateBound {
+		klog.V(2).InfoS("skip reconcile clusterResourceBinding that is not bound", "state", resourceBinding.Spec.State, "resource binding", bindingRef)
+		return ctrl.Result{}, nil
+	}
+
+	// generate the works
+	return r.syncAllWork(ctx, resourceBinding.DeepCopy())
+}
+
+// handleDelete handle a deleting binding
+func (r *Reconciler) handleDelete(ctx context.Context, resourceBinding *fleetv1beta1.ClusterResourceBinding) (ctrl.Result, error) {
+	klog.V(4).InfoS("handle deleting resource binding", "resource binding", klog.KObj(resourceBinding))
+	// list all the corresponding works if exist
+	works, err := r.listAllWorksAssociated(ctx, resourceBinding)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	// remove the overrider finalizer on the binding if all the work is deleted
+	if len(works) == 0 {
+		controllerutil.RemoveFinalizer(resourceBinding, fleetv1beta1.WorkFinalizer)
+		if err = r.Client.Update(ctx, resourceBinding); err != nil {
+			klog.ErrorS(err, "failed to remove the work finalizer from resource binding", "resource binding", klog.KObj(resourceBinding))
+			return ctrl.Result{}, controller.NewAPIServerError(false, err)
+		}
+	}
+	klog.V(4).InfoS("the resource binding still has undeleted work", "resource binding", klog.KObj(resourceBinding),
+		"number of associated work", len(works))
+	return ctrl.Result{RequeueAfter: 30 * time.Second}, nil
```

Comment:
we've already watched the work delete event. We requeue the request for backup? 5 min is good enough?
Could you please add a note there?